x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Phased.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.plan.logical;
+
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.xpack.esql.analysis.Analyzer;
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+import org.elasticsearch.xpack.esql.core.util.Holder;
+
+import java.util.List;
+
+/**
+ * Marks a {@link LogicalPlan} node as requiring multiple ESQL executions to run.
+ * All logical plans are now run by:
+ * <ol>
+ *     <li>{@link Analyzer analyzing} the entire query</li>
+ *     <li>{@link Phased#extractFirstPhase extracting} the first phase from the
+ *         logical plan</li>
+ *     <li>if there isn't a first phase, run the entire logical plan and return the
+ *         results. you are done.</li>
+ *     <li>if there is first phase, run that</li>
+ *     <li>{@link Phased#applyResultsFromFirstPhase appling} the results from the
+ *         first phase into the logical plan</li>
+ *     <li>start over from step 2 using the new logical plan</li>
```

Comment:
Could we add an example of how this is intended to work, and how results are applied?

Because, the same description can lead to 2 very different ways of execution:
```
FROM idx
\_INLINESTATS[m=max(x) by z]
  \_INLINESTATS[avg(m)]
```
We could split this at the top
```
1. phase:
FROM idx
\_STATS[m=max(x) by z]

2. phase:
FROM idx
\_HASH_JOIN[local relation computed above]
  \_INLINESTATS[avg(m)]
-> will be split into another 2 phases
```
or we could split this at the bottom
```
1. phase:
FROM idx
\_INLINESTATS[m=max(x) by z]
  \_STATS[avg(m)]
-> needs to be split again

2. phase:
FROM idx
\_INLINESTATS[m=max(x) by z]
  \_HASHJOIN[local relation computed above]
-> also needs to be split again
```
The second way is clearly bad, but can be mistakenly assumed (I did!) :D